lib/govuk_navigation_helpers/curated_taxonomy_sidebar_links.rb

```diff
@@ -0,0 +1,93 @@
+require 'set'
+
+module GovukNavigationHelpers
+  class CuratedTaxonomySidebarLinks
+    def initialize(content_item)
+      @content_item = content_item
+    end
+
+    def related_items
+      @related_items ||=
+        taxon_links +
+        elsewhere_on_gov_uk_links +
+        elsewhere_on_the_web_links
+    end
+
+  private
+
+    def taxon_links
+      @content_item.parent_taxons.map do |taxon|
+        {
+          title: taxon.title,
+          url: taxon.base_path,
+          description: taxon.description,
+          related_content: related_content_by_taxon[taxon],
+        }
+      end
+    end
+
+    def elsewhere_on_gov_uk_links
+      elsewhere_items = related_content_by_taxon[nil]
+      return [] if elsewhere_items.empty?
+
+      [
+        {
+          title: 'Elsewhere on GOV.UK',
+          related_content: related_content_by_taxon[nil],
+        },
+      ]
+    end
+
+    def elsewhere_on_the_web_links
+      return [] if @content_item.external_links.empty?
+
+      external_links = @content_item.external_links.map do |link|
+        {
+          title: link['title'],
+          link: link['url'],
+        }
+      end
+
+      [
+        {
+          title: 'Elsewhere on the web',
+          related_content: external_links,
+        },
+      ]
+    end
+
+    def related_content_by_taxon
+      return @related_items_by_taxon unless @related_items_by_taxon.nil?
+
+      curated_related_items = @content_item.curated_taxonomy_sidebar_links.to_set
+      items_by_taxon = {}
+
+      @content_item.parent_taxons.each do |taxon|
+        items_related_to_taxon = filter_items_by_taxon(curated_related_items, taxon)
+        items_by_taxon[taxon] = format_for_sidebar(items_related_to_taxon)
+        curated_related_items = undisplayed_items(curated_related_items, items_related_to_taxon)
+      end
+
+      items_by_taxon[nil] = format_for_sidebar(curated_related_items)
```

Comment:
I understand now, but it created a bit of confusion. Do you think we could re-write this?